devdoc/waiter_queue_requirements.md

```diff
@@ -0,0 +1,69 @@
+# `waiter_queue` requirements
+
+## Overview
+
+`waiter_queue` is a module that allows a consumers to park calls and for producers to execute them when data becomes available. It has two parts: `waiter_queue_ll` that contains the logic for parking and executing calls and `waiter_queue_hl` that contains logic for making `waiter_queue` thread-safe.
+
+
+## Exposed API
+
+### waiter_queue_ll
+```
+#define WAITER_QUEUE_CALL_REASON_VALUES \
+    WAITER_QUEUE_CALL_REASON_POPPED, \
+    WAITER_QUEUE_CALL_REASON_ABANDONED
+
+MU_DEFINE_ENUM(WAITER_QUEUE_CALL_REASON, WAITER_QUEUE_CALL_REASON_VALUES);
+
+typedef struct WAITER_QUEUE_LL_TAG* WAITER_QUEUE_LL_HANDLE;
+typedef bool(*POP_CALLBACK)(void* context, void* data, bool* continue_processing, WAITER_QUEUE_CALL_REASON reason);
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+    MOCKABLE_FUNCTION(, WAITER_QUEUE_LL_HANDLE, waiter_queue_ll_create);
+    MOCKABLE_FUNCTION(, void, waiter_queue_ll_destroy, WAITER_QUEUE_LL_HANDLE, waiter_queue_ll);
+
+    MOCKABLE_FUNCTION(, int, waiter_queue_ll_push, WAITER_QUEUE_LL_HANDLE, waiter_queue_ll, POP_CALLBACK, pop_callback, void*, pop_callback_context);
+    MOCKABLE_FUNCTION(, int, waiter_queue_ll_pop, WAITER_QUEUE_LL_HANDLE, waiter_queue_ll, void*, data);
+    MOCKABLE_FUNCTION(, void, waiter_queue_ll_abandon, WAITER_QUEUE_LL_HANDLE, waiter_queue_ll);
```

Comment:
>waiter_queue_ll_abandon [](http://example.com/codeflow?start=30&length=23)

What does this do? I don't recall discussing it and I don't see it in the doc.

Reply:
This is needed to abandon all parked calls when `waiter_queue_hl_destroy` is called. `waiter_queue_hl` needs some way of telling `waiter_queue_ll` to abandon the calls.